### ISSUE TITLE
panic: “Remove only what we have; qed”

### ISSUE THREAD

**Issue by h4x3rotab** [Phala-Network/phala-blockchain#527]
Logs from collators:

```
Oct 13 03:16:54 khala-genesis-collators collator_1.sh[388699]: 2021-10-13 03:16:54 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 05:41:06 khala-genesis-collators collator_1.sh[388699]: 2021-10-13 05:41:06 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 07:09:12 khala-genesis-collators collator_1.sh[388699]: 2021-10-13 07:09:12 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 07:19:00 khala-genesis-collators collator_1.sh[388699]: 2021-10-13 07:19:00 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
```

```
Oct 13 03:17:06 khala-genesis-collators collator_2.sh[395208]: 2021-10-13 03:17:06 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 05:41:18 khala-genesis-collators collator_2.sh[395208]: 2021-10-13 05:41:18 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 07:09:24 khala-genesis-collators collator_2.sh[395208]: 2021-10-13 07:09:24 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 07:19:12 khala-genesis-collators collator_2.sh[395208]: 2021-10-13 07:19:12 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
```

```
Oct 13 03:17:12 khala-genesis-collators collator_3.sh[402373]: 2021-10-13 03:17:12 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 05:41:30 khala-genesis-collators collator_3.sh[402373]: 2021-10-13 05:41:30 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 07:09:36 khala-genesis-collators collator_3.sh[402373]: 2021-10-13 07:09:36 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
Oct 13 07:19:24 khala-genesis-collators collator_3.sh[402373]: 2021-10-13 07:19:24 [Parachain] panicked at 'Remove only what we have; qed.', /home/h4x/workspace/khala-parachain/pallets/phala/src/stakepool.rs:854:26
```

**Comment by h4x3rotab** [comment]
It turned out a fixed point precision issue

**Comment by h4x3rotab** [comment]
Turned out not fully fixed. Reopened with sample:

- Pool: 386
- 43E9fDcG5mFxRRxW2dR4NDU3YPmQfGPsjD3tzQWnN29h8M2A 0xec2b3a4dbe2588c117d2fef9c0f3cfd18b41b9395d83723f4d4c45591a1af35c
- 43E9fDcZLuybK4AXLrrhgLZfL4bjiERccXW3tcJXAS6KUTkG 0xd47e25c8cbc8e6e24f03289563baab06ac57e0d30572e448d5243c3afdb4b479

**Comment by h4x3rotab** [comment]
I just found out the root cause of the problem. The fix we have applied at https://github.com/Phala-Network/phala-blockchain/commit/9c798d1a009b436303acc624120a9f9566163776 has patched the hole of the precision problem.

The remaining problem lies in the dirty withdraw request. There are two pools affected:

- Pool 726, request from user `44uyiTRJ2gEpHYcBzn3TjXhkFqdy3Bbbt742Gyx85mt2KfDg` withdrawing 3000 PHA
- Pool 385, request from user `44ne2URJTvTJrZT7yn2HCPAvEhjak1hTAknRjJiskTRjycnC` withdrawing 5533 PHA

Behind the both requests, the users actually don't hold any share of the stake pool. Therefore it violates the assumption of `PoolInfo::remove_stake` that the requesting stake mustn't exceed the total stake a user holds. This problem was addressed in #490.

Before the fix to #490, a user can withdraw multiple times, and there's no proper check against the total requested withdraw amounts. The fix was to only allow a single withdraw request in the queue, and it runs a storage migration to remove all the existing duplicated requests in the pools. However there's an edge case: if a user has two withdraw requests, and the first one was fulfilled before the #490 runtime upgrade, the remaining one was not removed by the storage migration. This leaves some dirty withdraw requests in the queues without any stake backed by the user. And when there's any free stake (no matter from contribution or reclaim), the stake pool will try to fulfill the dirty requests, resulting a panic.

In #585, a script to examine the dirty requests in the withdraw queues is merged ([b527-check-bad-req.js](https://github.com/Phala-Network/phala-blockchain/blob/0b03596a03e0aac381751d099ae94168ad1987af/scripts/js/bugfix-tools/b527-check-bad-req.js)). Fortunately only two pools are affected. The next step, we can add a backfill function in the next runtime upgrade to remove the two bad requests.
